lib/checks/lists/invalid-children-evaluate.js

```diff
@@ -0,0 +1,75 @@
+import { isVisibleToScreenReaders } from '../../commons/dom';
+import { getExplicitRole } from '../../commons/aria';
+
+export default function invalidChildrenEvaluate(
+  node,
+  options = {},
+  virtualNode
+) {
+  const relatedNodes = [];
+  const issues = [];
+  if (!virtualNode.children) {
+    return undefined;
+  }
+
+  const vChildren = mapWithNested(virtualNode.children);
+  while (vChildren.length) {
+    const { vChild, nested } = vChildren.shift();
+    if (options.divGroups && !nested && isDivGroup(vChild)) {
+      if (!vChild.children) {
+        return undefined;
+      }
+      const vGrandChildren = mapWithNested(vChild.children, true);
+      vChildren.push(...vGrandChildren);
+      continue;
+    }
+
+    const issue = isInvalidChild(vChild, nested, options);
+    if (!issue) {
+      continue;
+    }
+    if (!issues.includes(issue)) {
+      issues.push(issue);
+    }
+    if (vChild?.actualNode?.nodeType === 1) {
+      relatedNodes.push(vChild.actualNode);
+    }
+  }
+  if (issues.length === 0) {
+    return false;
+  }
+
+  this.data({ values: issues.join(', ') });
+  this.relatedNodes(relatedNodes);
+  return true;
+}
+
+function isInvalidChild(
+  vChild,
+  nested,
+  { validRoles = [], validNodeNames = [] }
+) {
+  const { nodeName, nodeType, nodeValue } = vChild.props;
+  const selector = nested ? 'div > ' : '';
+  if (nodeType === 3 && nodeValue.trim() !== '') {
+    return selector + `#text`;
+  }
+  if (nodeType !== 1 || !isVisibleToScreenReaders(vChild)) {
+    return false;
+  }
+
+  const role = getExplicitRole(vChild);
```

Comment:
We should we use `getRole` instead to account for role conflict. Something like this would fail even though each is treated as a listitem (this currently does not fail today)

```html
<ul>
  <li role="none" aria-label="hello"></li>
  <li role="none" tabindex="0">Hello</li>
</ul>
```

Reply:
I don't think people should rely on fallback roles to pass. We're failing this after all. The one thing I do think is worth considering is to allow role=none on empty elements. We don't allow that currently, so I think that's a separate PR.

One thing I'd like your input on is in how this rule is becoming stricter. Something like the following was passing before. I don't think it should have been, but I'm not sure if we need to more explicitly call it out in the PR title, or even create a separate PR so its separated in the changelog. WDYT?

```html
<!-- This passes today, because it's an li and we then don't look at the role -->
<ul><li role="button">Button!</li></ul>
```